File: src/backend/internal/model/token_usage.go
package model

import "time"

// TokenUsage 仅携带原生计量。输入已包含缓存；nil 表示未上报，不能等同零。
// Complete 仅表示本次派发主循环统计完整，不代表账户账单或所有子代理。
type TokenUsage struct {
	Provider            string    `json:"provider"`
	Source              string    `json:"source"`
	Model               string    `json:"model,omitempty"`
	InputTokens         *int64    `json:"input_tokens,omitempty"`
	OutputTokens        *int64    `json:"output_tokens,omitempty"`
	CacheReadTokens     *int64    `json:"cache_read_tokens,omitempty"`
	CacheWriteTokens    *int64    `json:"cache_write_tokens,omitempty"`
	ReasoningTokens     *int64    `json:"reasoning_tokens,omitempty"`
	ContextTokens       *int64    `json:"context_tokens,omitempty"`
	ContextWindowTokens *int64    `json:"context_window_tokens,omitempty"`
	Complete            bool      `json:"complete"`
	ObservedAt          time.Time `json:"observed_at"`
}

// Valid 限制为 JS 安全整数并拒绝互相矛盾的细分值。
func (u *TokenUsage) Valid() bool {
	if u == nil || u.Source != "actual" || (u.Provider != "codex" && u.Provider != "claude") || u.ObservedAt.IsZero() {
		return false
	}
	for _, n := range []*int64{u.InputTokens, u.OutputTokens, u.CacheReadTokens, u.CacheWriteTokens, u.ReasoningTokens, u.ContextTokens, u.ContextWindowTokens} {
		if n != nil && (*n < 0 || *n > 9007199254740991) {
			return false
		}
	}
	if u.InputTokens != nil {
		var cached int64
		if u.CacheReadTokens != nil {
			cached += *u.CacheReadTokens
		}
		if u.CacheWriteTokens != nil {
			cached += *u.CacheWriteTokens
		}
		if cached > *u.InputTokens {
			return false
		}
	}
	if u.ReasoningTokens != nil && u.OutputTokens != nil && *u.ReasoningTokens > *u.OutputTokens {
		return false
	}
	return !u.Complete || (u.InputTokens != nil && u.OutputTokens != nil)
}

File: src/frontend/src/types/tokenUsage.ts
/** 原生输入含缓存；缓存、推理分别是输入、输出的子集。缺省表示未知。 */
export interface TokenUsage {
  context_events?: { id: string; kind: string; status: string; started_at?: string; ended_at?: string }[];
  provider: string;
  source: 'actual';
  model?: string;
  input_tokens?: number;
  output_tokens?: number;
  cache_read_tokens?: number;
  cache_write_tokens?: number;
  reasoning_tokens?: number;
  context_tokens?: number;
  context_window_tokens?: number;
  complete: boolean;
  observed_at?: string;
}

export function isTokenUsage(value: unknown): value is TokenUsage {
  if (!value || typeof value !== 'object') return false;
  const u = value as Partial<TokenUsage>;
  if (u.source !== 'actual' || !['codex', 'claude'].includes(u.provider ?? '') || !u.observed_at || !Number.isFinite(Date.parse(u.observed_at))) return false;
  if (u.context_events != null && (!Array.isArray(u.context_events) || u.context_events.length > 64 || u.context_events.some(e => !e || typeof e.id !== 'string' || !e.id || e.id.length > 200 || e.kind !== 'compaction' || !['running','complete'].includes(e.status) || [e.started_at,e.ended_at].some(t => t != null && !Number.isFinite(Date.parse(t)))))) return false;
  const counts = [u.input_tokens, u.output_tokens, u.cache_read_tokens, u.cache_write_tokens, u.reasoning_tokens, u.context_tokens, u.context_window_tokens];
  if (counts.some((n) => n != null && (!Number.isSafeInteger(n) || n < 0))) return false;
  if (u.input_tokens != null && (u.cache_read_tokens ?? 0) + (u.cache_write_tokens ?? 0) > u.input_tokens) return false;
  if (u.output_tokens != null && u.reasoning_tokens != null && u.reasoning_tokens > u.output_tokens) return false;
  return !u.complete || (u.input_tokens != null && u.output_tokens != null);
}
